File: EnglishPhoneticApp/EnglishPhoneticApp/EnglishPhoneticApp/EnglishPhoneticApp/ViewModels/TextbookViewModel.swift
import Foundation
import SwiftUI
import Combine

class TextbookViewModel: ObservableObject {
    @Published var textbooks: [Textbook] = []
    @Published var isProcessing = false
    @Published var processingProgress: String = ""
    
    private let dataStore = DataStoreService.shared
    private let ocrService = OCRService()
    private let phoneticService = PhoneticDictionaryService.shared
    
    init() {
        loadTextbooks()
        setupSamplePages()
    }
    
    func loadTextbooks() {
        textbooks = dataStore.loadTextbooks()
    }
    
    private func saveTextbooks() {
        dataStore.saveTextbooks(textbooks)
    }
    
    private func setupSamplePages() {
        guard !textbooks.contains(where: { $0.name == "示例课本" }) else { return }
        
        let sampleUnit = TextbookUnit(name: "Unit 1 Hello", order: 0)
        let sampleTextbook = Textbook(
            name: "示例课本",
            coverImagePath: nil,
            units: [sampleUnit]
        )
        
        textbooks.insert(sampleTextbook, at: 0)
        saveTextbooks()
    }
    
    func addTextbook(name: String) -> Textbook {
        let textbook = Textbook(name: name)
        textbooks.append(textbook)
        saveTextbooks()
        return textbook
    }
    
    func deleteTextbook(_ textbook: Textbook) {
        for unit in textbook.units {
            for page in unit.pages {
                dataStore.deleteImage(filename: page.imagePath)
            }
        }
        
        textbooks.removeAll { $0.id == textbook.id }
        saveTextbooks()
    }
    
    func addPage(to textbookId: UUID, unitId: UUID, image: UIImage, pageName: String, completion: @escaping (TextbookPage?) -> Void) {
        isProcessing = true
        processingProgress = "正在识别文字..."
        
        let pageId = UUID()
        
        guard let imageFilename = dataStore.saveImage(image, forPageId: pageId) else {
            isProcessing = false
            completion(nil)
            return
        }
        
        ocrService.recognizeText(in: image) { [weak self] results in
            guard let self = self else {
                completion(nil)
                return
            }
            
            self.processingProgress = "正在生成音标..."
            
            var annotations: [WordAnnotation] = []
            for result in results {
                let phonetic = self.phoneticService.lookup(result.text) ?? ""
                let annotation = WordAnnotation(
                    word: result.text,
                    phonetic: phonetic,
                    normalizedX: Double(result.boundingBox.origin.x),
                    normalizedY: Double(result.boundingBox.origin.y),
                    normalizedWidth: Double(result.boundingBox.width),
                    normalizedHeight: Double(result.boundingBox.height)
                )
                annotations.append(annotation)
            }
            
            let page = TextbookPage(
                id: pageId,
                name: pageName,
                imagePath: imageFilename,
                order: 0,
                annotations: annotations
            )
            
            if let textbookIndex = self.textbooks.firstIndex(where: { $0.id == textbookId }),
               let unitIndex = self.textbooks[textbookIndex].units.firstIndex(where: { $0.id == unitId }) {
                
                self.textbooks[textbookIndex].units[unitIndex].pages.append(page)
                self.saveTextbooks()
            }
            
            self.isProcessing = false
            self.processingProgress = ""
            
            DispatchQueue.main.async {
                completion(page)
            }
        }
    }
    
    func updateAnnotations(for pageId: UUID, in textbookId: UUID, unitId: UUID, annotations: [WordAnnotation]) {
        guard let textbookIndex = textbooks.firstIndex(where: { $0.id == textbookId }),
              let unitIndex = textbooks[textbookIndex].units.firstIndex(where: { $0.id == unitId }),
              let pageIndex = textbooks[textbookIndex].units[unitIndex].pages.firstIndex(where: { $0.id == pageId }) else {
            return
        }
        
        textbooks[textbookIndex].units[unitIndex].pages[pageIndex].annotations = annotations
        saveTextbooks()
    }
    
    func deletePage(_ page: TextbookPage, from textbookId: UUID, unitId: UUID) {
        dataStore.deleteImage(filename: page.imagePath)
        
        guard let textbookIndex = textbooks.firstIndex(where: { $0.id == textbookId }),
              let unitIndex = textbooks[textbookIndex].units.firstIndex(where: { $0.id == unitId }) else {
            return
        }
        
        textbooks[textbookIndex].units[unitIndex].pages.removeAll { $0.id == page.id }
        saveTextbooks()
    }
    
    func renamePage(_ page: TextbookPage, to newName: String, in textbookId: UUID, unitId: UUID) {
        guard let textbookIndex = textbooks.firstIndex(where: { $0.id == textbookId }),
              let unitIndex = textbooks[textbookIndex].units.firstIndex(where: { $0.id == unitId }),
              let pageIndex = textbooks[textbookIndex].units[unitIndex].pages.firstIndex(where: { $0.id == page.id }) else {
            return
        }
        
        textbooks[textbookIndex].units[unitIndex].pages[pageIndex].name = newName
        saveTextbooks()
        
        // 触发 Published 更新，确保视图刷新
        objectWillChange.send()
    }
    
    func addUnit(to textbookId: UUID, name: String) -> TextbookUnit? {
        guard let textbookIndex = textbooks.firstIndex(where: { $0.id == textbookId }) else {
            return nil
        }
        
        let order = textbooks[textbookIndex].units.count
        let unit = TextbookUnit(name: name, order: order)
        textbooks[textbookIndex].units.append(unit)
        saveTextbooks()
        return unit
    }
}

File: androidVersion/app/src/main/java/com/liudong/bookread/viewmodel/TextbookViewModel.kt
package com.liudong.bookread.viewmodel

import android.app.Application
import android.graphics.Bitmap
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.liudong.bookread.model.Textbook
import com.liudong.bookread.model.TextbookPage
import com.liudong.bookread.model.TextbookUnit
import com.liudong.bookread.model.WordAnnotation
import com.liudong.bookread.service.DataStoreService
import com.liudong.bookread.service.OCRService
import com.liudong.bookread.service.PhoneticDictionaryService
import com.liudong.bookread.service.SpeechService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import java.util.UUID

class TextbookViewModel(application: Application) : AndroidViewModel(application) {
    private val app = getApplication<Application>()
    private val context get() = app.applicationContext

    private val _textbooks = MutableStateFlow<List<Textbook>>(emptyList())
    val textbooks: StateFlow<List<Textbook>> = _textbooks.asStateFlow()

    private val _isProcessing = MutableStateFlow(false)
    val isProcessing: StateFlow<Boolean> = _isProcessing.asStateFlow()

    private val _processingProgress = MutableStateFlow("")
    val processingProgress: StateFlow<String> = _processingProgress.asStateFlow()

    init {
        viewModelScope.launch {
            loadTextbooks()
            setupSamplePages()
        }
        PhoneticDictionaryService.loadDictionary(context)
        SpeechService.initialize(context)
    }

    private suspend fun loadTextbooks() {
        _textbooks.value = DataStoreService.loadTextbooks(context)
    }

    private fun saveTextbooks() {
        viewModelScope.launch {
            DataStoreService.saveTextbooks(context, _textbooks.value)
        }
    }

    private fun setupSamplePages() {
        if (_textbooks.value.any { it.name == "示例课本" }) return
        val sampleUnit = TextbookUnit(name = "Unit 1 Hello", order = 0)
        val sampleTextbook = Textbook(name = "示例课本", units = listOf(sampleUnit))
        _textbooks.value = listOf(sampleTextbook) + _textbooks.value
        saveTextbooks()
    }

    fun addTextbook(name: String): Textbook {
        val textbook = Textbook(name = name)
        _textbooks.value = _textbooks.value + textbook
        saveTextbooks()
        return textbook
    }

    fun deleteTextbook(textbook: Textbook) {
        textbook.units.flatMap { it.pages }.forEach { page ->
            DataStoreService.deleteImage(context, page.imagePath)
        }
        _textbooks.value = _textbooks.value.filter { it.id != textbook.id }
        saveTextbooks()
    }

    fun addPage(
        textbookId: String,
        unitId: String,
        bitmap: Bitmap,
        pageName: String,
        onComplete: (TextbookPage?) -> Unit
    ) {
        viewModelScope.launch {
            _isProcessing.value = true
            _processingProgress.value = "正在识别文字..."

            val pageId = UUID.randomUUID().toString()
            val imageFilename = DataStoreService.saveImage(context, bitmap, pageId)
                ?: run {
                    _isProcessing.value = false
                    onComplete(null)
                    return@launch
                }

            val ocrResults = OCRService.recognizeText(bitmap).first()

            _processingProgress.value = "正在生成音标..."
            val annotations = ocrResults.map { result ->
                WordAnnotation(
                    word = result.text,
                    phonetic = PhoneticDictionaryService.lookup(result.text) ?: "",
                    normalizedX = result.boundingBox.left.toDouble() / bitmap.width,
                    normalizedY = 1.0 - (result.boundingBox.bottom.toDouble() / bitmap.height),
                    normalizedWidth = result.boundingBox.width().toDouble() / bitmap.width,
                    normalizedHeight = result.boundingBox.height().toDouble() / bitmap.height
                )
            }

            val page = TextbookPage(
                id = pageId,
                name = pageName,
                imagePath = imageFilename,
                order = 0,
                annotations = annotations
            )

            val updated = _textbooks.value.map { tb ->
                if (tb.id == textbookId) {
                    tb.copy(units = tb.units.map { unit ->
                        if (unit.id == unitId) unit.copy(pages = unit.pages + page) else unit
                    })
                } else tb
            }
            _textbooks.value = updated
            saveTextbooks()

            _isProcessing.value = false
            _processingProgress.value = ""
            onComplete(page)
        }
    }

    fun updateAnnotations(textbookId: String, unitId: String, pageId: String, annotations: List<WordAnnotation>) {
        val updated = _textbooks.value.map { tb ->
            if (tb.id == textbookId) {
                tb.copy(units = tb.units.map { unit ->
                    if (unit.id == unitId) {
                        unit.copy(pages = unit.pages.map { p ->
                            if (p.id == pageId) p.copy(annotations = annotations) else p
                        })
                    } else unit
                })
            } else tb
        }
        _textbooks.value = updated
        saveTextbooks()
    }

    fun deletePage(textbookId: String, unitId: String, page: TextbookPage) {
        DataStoreService.deleteImage(context, page.imagePath)
        val updated = _textbooks.value.map { tb ->
            if (tb.id == textbookId) {
                tb.copy(units = tb.units.map { unit ->
                    if (unit.id == unitId) unit.copy(pages = unit.pages.filter { it.id != page.id })
                    else unit
                })
            } else tb
        }
        _textbooks.value = updated
        saveTextbooks()
    }

    fun renamePage(textbookId: String, unitId: String, page: TextbookPage, newName: String) {
        val updated = _textbooks.value.map { tb ->
            if (tb.id == textbookId) {
                tb.copy(units = tb.units.map { unit ->
                    if (unit.id == unitId) {
                        unit.copy(pages = unit.pages.map { p ->
                            if (p.id == page.id) p.copy(name = newName) else p
                        })
                    } else unit
                })
            } else tb
        }
        _textbooks.value = updated
        saveTextbooks()
    }

    fun addUnit(textbookId: String, name: String): TextbookUnit? {
        val textbook = _textbooks.value.find { it.id == textbookId } ?: return null
        val order = textbook.units.size
        val unit = TextbookUnit(name = name, order = order)
        val updated = _textbooks.value.map { tb ->
            if (tb.id == textbookId) tb.copy(units = tb.units + unit) else tb
        }
        _textbooks.value = updated
        saveTextbooks()
        return unit
    }

    fun reprocessPage(textbookId: String, unitId: String, page: TextbookPage, bitmap: Bitmap, onComplete: (List<WordAnnotation>) -> Unit) {
        viewModelScope.launch {
            _isProcessing.value = true
            _processingProgress.value = "重新识别中..."

            val ocrResults = OCRService.recognizeText(bitmap).first()
            val annotations = ocrResults.map { result ->
                WordAnnotation(
                    word = result.text,
                    phonetic = PhoneticDictionaryService.lookup(result.text) ?: "",
                    normalizedX = result.boundingBox.left.toDouble() / bitmap.width,
                    normalizedY = 1.0 - (result.boundingBox.bottom.toDouble() / bitmap.height),
                    normalizedWidth = result.boundingBox.width().toDouble() / bitmap.width,
                    normalizedHeight = result.boundingBox.height().toDouble() / bitmap.height
                )
            }

            val updated = _textbooks.value.map { tb ->
                if (tb.id == textbookId) {
                    tb.copy(units = tb.units.map { u ->
                        if (u.id == unitId) {
                            u.copy(pages = u.pages.map { p ->
                                if (p.id == page.id) p.copy(annotations = annotations) else p
                            })
                        } else u
                    })
                } else tb
            }
            _textbooks.value = updated
            saveTextbooks()

            _isProcessing.value = false
            _processingProgress.value = ""
            onComplete(annotations)
        }
    }
}
